akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/DDataRememberEntitiesCoordinatorStore.scala

```diff
@@ -0,0 +1,122 @@
+import akka.cluster.ddata.SelfUniqueAddress
+import akka.cluster.sharding.ClusterShardingSettings
+import akka.cluster.sharding.ShardRegion.ShardId
+
+private[akka] object DDataRememberEntitiesCoordinatorStore {
+  def props(typeName: String, settings: ClusterShardingSettings, replicator: ActorRef, majorityMinCap: Int): Props =
+    Props(new DDataRememberEntitiesCoordinatorStore(typeName, settings, replicator, majorityMinCap))
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] class DDataRememberEntitiesCoordinatorStore(
+    typeName: String,
+    settings: ClusterShardingSettings,
+    replicator: ActorRef,
+    majorityMinCap: Int)
+    extends Actor
+    with Stash
+    with ActorLogging {
+
+  implicit val node: Cluster = Cluster(context.system)
+  implicit val selfUniqueAddress: SelfUniqueAddress = SelfUniqueAddress(node.selfUniqueAddress)
+
+  private val readMajority = ReadMajority(settings.tuningParameters.waitingForStateTimeout, majorityMinCap)
+  private val writeMajority = WriteMajority(settings.tuningParameters.updatingStateTimeout, majorityMinCap)
+
+  private val AllShardsKey = GSetKey[String](s"shard-${typeName}-all")
+  private var allShards = Set.empty[ShardId]
+
+  // eager load of remembered shard ids
+  def getAllShards(): Unit = {
+    replicator ! Replicator.Get(AllShardsKey, readMajority)
+  }
+  getAllShards()
+
+  override def receive: Receive = waitingForRead()
+
+  def waitingForRead(): Receive = {
+
+    case RememberEntitiesCoordinatorStore.GetShards =>
+      stash()
+
+    case g @ Replicator.GetSuccess(AllShardsKey, _) =>
+      allShards = g.get(AllShardsKey).elements
+      context.become(idle())
+      unstashAll()
+
+    case Replicator.GetFailure(AllShardsKey, _) =>
+      log.error(
+        "The ShardCoordinator was unable to get all shards state within 'waiting-for-state-timeout': {} millis (retrying)",
+        readMajority.timeout.toMillis)
+      // repeat until GetSuccess
+      getAllShards()
+
+    case Replicator.NotFound(AllShardsKey, _) =>
+      allShards = Set.empty
+      context.become(idle())
+      unstashAll()
+  }
+
+  def idle(): Receive = {
+    case RememberEntitiesCoordinatorStore.GetShards =>
+      // FIXME should we clear the set once we have replied given that we know there is only one get-request ever?
+      sender ! RememberEntitiesCoordinatorStore.RememberedShards(allShards)
+    case RememberEntitiesCoordinatorStore.AddShard(shardId) =>
+      if (!allShards.contains(shardId)) {
+        addShard(shardId, sender())
+      } else {
+        // already known started shard
+        sender() ! RememberEntitiesCoordinatorStore.UpdateDone(shardId)
+      }
+  }
+
+  def addShard(newShard: ShardId, replyTo: ActorRef): Unit = {
+    replicator ! Replicator.Update(AllShardsKey, GSet.empty[String], writeMajority, Some(newShard))(_ + newShard)
+    context.become(waitingForWrite(replyTo))
+  }
+
+  def waitingForWrite(replyTo: ActorRef): Receive = {
```

Comment:
AddShard not handled here. I think we can simplify and make it more robust by tunneling the `sender()` in the `Replicator.Update.request field. Then it will come back in UpdateSuccess/UpdateTimeout and we don't need to keep track of it in the state of the actor.

Since retries are handled by the coordinator that will simplify things here. We don't need the `idle` and `waitingForWrite` behaviors. It can be one `initializing` where it's waiting for the initial GetSuccess and then one `active` that handles everything after that.

Reply:
I took it one step further to a single receive, much cleaner and easier to read 👍